shared/naturalcrit/codeEditor/codeEditor.jsx

```diff
@@ -57,6 +66,12 @@ const CodeEditor = createClass({
 		} else if(this.codeMirror?.getValue() != this.props.value) { //update editor contents if brew.text is changed from outside
 			this.codeMirror.setValue(this.props.value);
 		}
+
+		setTimeout(()=>{
+			snapshot.forEach((fold)=>{
+				this.codeMirror.foldCode(fold, { scanUp: false }, 'fold');
+			});
+		}, 0);
```

Comment:
I agree with you, this seems like a hacky way to do this. Can you explain again your findings before you tried this method out? I know you mentioned it in Gitter but here we can have it in a note that we can refer to as we keep investigating.

Reply:
Absolutely! I did a lot of debugging around this, opening console logs in the various lifecycle hooks to catch where it ends up unfolding the code. The code unfolds on the dom after `componentDidUpdate` completes, which is seemingly why the timeout (even though it's set to 0ms) successfully refolds the code that didn't move.

An important note: removing the `this.codeMirror.on('change', (cm)=>{this.props.onChange(cm.getValue());});` event did prevent the code from being unfolded on a change during my testing, so the issue likely lies within how React interacts with CodeMirror and not in CodeMirror itself.

Whatever is unfolding the code does result in a CodeMirror `unfold` event being fired. You can use `this.codemirror.on('unfold', (cm)=>{console.log('unfold');});` to see that.

Reply:
Are you able to run through the debugger in Chrome to locate exactly when the code unfolds? I think you are right, this sounds like an unintended consequence of something we did interacting with React.

Reply:
Yeah my next plan with it was to step over in the debugger until it finally unfolds (I haven't had much time to keep working on it in the last few days). I'm concerned this is going to be something that's not necessarily a result of how CodeMirror was implemented with React, but that it might be a limitation imposed onto CodeMirror by React itself and how it interacts with the dom. No evidence to support that, just a thought that keeps running through my head.